mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/BlockMatrix.scala

```diff
@@ -172,6 +172,64 @@ class BlockMatrix(
     assert(cols <= nCols, s"The number of columns $cols is more than claimed $nCols.")
   }
 
+  def validate: Unit = {
+    logDebug("Validating BlockMatrix...")
+    // check if the matrix is larger than the claimed dimensions
```

Comment:
Maybe we can cache the block info as a private lazy val to avoid having multiple passes to the matrix data:

~~~
private lazy val blockInfo = blocks.mapValues(block => (block.numRows, block.numCols)).cache()
~~~